crates/db_schema/src/impls/actor_language.rs

```diff
@@ -293,25 +293,37 @@ impl CommunityLanguage {
       return Ok(());
     }
 
+    let form = lang_ids
+      .into_iter()
+      .map(|language_id| CommunityLanguageForm {
+        community_id: for_community_id,
+        language_id,
+      })
+      .collect::<Vec<_>>();
+
     conn
       .build_transaction()
       .run(|conn| {
         Box::pin(async move {
           use crate::schema::community_language::dsl::{community_id, community_language};
+          use diesel::result::DatabaseErrorKind::UniqueViolation;
           // Clear the current languages
           delete(community_language.filter(community_id.eq(for_community_id)))
             .execute(conn)
             .await?;
 
-          for l in lang_ids {
-            let form = CommunityLanguageForm {
-              community_id: for_community_id,
-              language_id: l,
-            };
-            insert_into(community_language)
-              .values(form)
-              .get_result::<Self>(conn)
-              .await?;
+          let insert_res = insert_into(community_language)
+            .values(form)
+            .get_result::<Self>(conn)
+            .await;
+
+          if let Err(Error::DatabaseError(UniqueViolation, _info)) = insert_res {
+            // race condition: this function was probably called simultaneously from another caller. ignore error
+            // tracing::warn!("unique error: {_info:#?}");
+            // _info.constraint_name() should be = "community_language_community_id_language_id_key"
+            return Ok(());
+          } else {
+            insert_res?;
```

Comment:
This fix seems like an ugly hack. Would be better to prevent the same community from being fetched/inserted simultaneously. I think it could be solved with db transactions https://github.com/LemmyNet/lemmy/issues/1161

Reply:
It's not pretty, but putting everything in a transaction I'm pretty sure will just make everything worse. It also doesn't automatically fix most things, as you can see by the DELETE and INSERT queries here already being in a transaction.

Reply:
I've posted in #1161 about two better solutions to this issue:

1. The code should ensure that ObjectId<Community>::dereference_from_http (or really any object) is only called once simultaneously. This could be done with a cache that maps object id to shared futures.
2. The code should ensure that all the work starting with `insert into community` should only happen once. this can be done by removing the on_conflict().do_update() in Community::create and if instead return the existing object if it exists. I don't know if this would break other code.


This PR only implements a lower-level fix. If (2) is implemented, this can be removed, but (1) and (2) are more complicated.

I'll leave this open, you can decide if you want to merge it or not.

Reply:
Also, I want to mention that the code base (imo) has a ton of these "hacks". Usually they are just far less visible because they are simply done using `if let Ok(x) = fallible_stuff {}` without an else clause, just silently throwing away all kinds of errors. I could have done the same here and it would be a single line of code.

Reply:
Yes some kind of cache or locking for dereference_from_http would make sense. The on_conflict().do_update() is necessary because Activitypub cant really distinguish between create and update actions, so we dont know which one is needed.